eventsocket/eventsocket.go

```diff
@@ -45,6 +49,63 @@ type Connection struct {
 	textreader    *textproto.Reader
 	err           chan error
 	cmd, api, evt chan *Event
+	hub           *RequestHub
+}
+
+// Request is a bgapi request struct
+type Request struct {
+	UUID      string
+	timestamp time.Time
+	resp      chan *Event
+	err       chan error
+	command   string
+}
+
+// RequestHub is the asyn request/reply hub for bgapi commands
+type RequestHub struct {
+	requests  map[string]*Request
+	inbound   chan *Request
+	responses chan *Event
+	timeout   chan bool
+}
+
+func (h *RequestHub) run() {
+	go func() {
+		for {
+			time.Sleep(1 * time.Second)
+			h.timeout <- true
+		}
+	}()
+	go func() {
+		for {
+			select {
+			case request := <-h.inbound:
+				h.requests[request.UUID] = request
+			case response := <-h.responses:
+				req, ok := h.requests[response.Header["Job-Uuid"].(string)]
+				if ok != true {
+					break
+				}
+				reply := response.Body
+				if strings.HasPrefix(reply, "-E") {
+					if len(reply) > 5 {
+						req.err <- errors.New(reply[5:])
+					} else {
+						req.err <- errors.New(reply)
+					}
+					break
+				}
+				req.resp <- response
+				delete(h.requests, req.UUID)
+			case _ = <-h.timeout:
+				for _, req := range h.requests {
```

Comment:
I don't love how the timeout is handled. It would be super interesting to see the inbound request have an embedded context, where it can set its own deadline. A simple implementation would be to setup a go routine when a request is received that checks the ctx.Done() channel and deletes itself from the requests map (in a concurrency safe way).